File: update_perf.py
import math
import time
import pymongo
import random

from pprint import pprint

NUM_DOCS = 125000
NUM_FIELDS = 100
NUM_UPDATES = 125000
NUM_FIELDS_UPDATED = 10

client = pymongo.MongoClient();
coll = client['test']['update_test']

def load():
    for i in range(NUM_DOCS):
        tmp = {"_id": i}
        for j in range(NUM_FIELDS):
            tmp[str(j)] = random.random()
        coll.insert(tmp);

def drive():
    for i in range(NUM_UPDATES):
        query = {"_id": i % NUM_DOCS}
        update = {}
        for j in range(NUM_FIELDS_UPDATED):
            randfield = int(math.floor(random.random() * NUM_FIELDS))
            update[str(randfield)] = random.random()
        update = {"$set": update}
        coll.update(query, update)

def main(): 
    coll.drop()
    load()
    starttime = time.time()
    drive()
    endtime = time.time()
    print "start time: " + str(starttime)
    print "end time: " + str(endtime)
    print "time elapsed: " + str(endtime - starttime)


if __name__ == "__main__":
    main()

File: update_perf.js
var NUM_DOCS = 100000;
var NUM_FIELDS = 100;
var NUM_UPDATES = 100000;
var NUM_FIELDS_UPDATED = 10;

var coll = db.update_perf_test;
coll.drop();
//first we load up test data

var tmp = {};
for(var i=0;i<NUM_DOCS;i++) {
    tmp = {_id: i};
    for(var j=0; j<NUM_FIELDS; j++) {
        tmp[j.toString()] = Math.random();
    }
    coll.insert(tmp);
}

//now we roughly test performance of updates
var query = {};
var update = {};
var randField = 0;
var startt = Date.now();
for(i=0; i<NUM_UPDATES; i++) {
    query = {_id: (i % NUM_DOCS)}
    update = {}
    for(j=0; j<NUM_FIELDS_UPDATED; j++) {
        randField = Math.random() * NUM_FIELDS;
        update[randField.toString()] = Math.random();
    }
    update = {$set: update};
    coll.update(query, update);
    //db.getLastErrorObj();
}
var endt = Date.now();
print("start: " + startt);
print("end: " + endt);
var totalTime = endt - startt;
print("time elapsed: " + totalTime);
